### DISCUSSION TITLE
Can I get canonical SMARTS?

### DISCUSSION THREAD

**Discussion by HelloJocelynLu** [rdkit/rdkit#6655] [Q&A] [answered]
Hi there,
I am trying to do molecular fragmentation, then collect fragment type based on their skeleton. Something like this:
<img width="600" alt="image" src="https://github.com/rdkit/rdkit/assets/34988596/a0a6cb00-3cd2-4a3c-91e0-ff657c6cbf12">
I want to define the skeleton based only on topology and aromaticity. I used a function like this:
```python3
from itertools import combinations
aromaticA = Chem.MolFromSmarts('[a]').GetAtomWithIdx(0)
aphalicA = Chem.MolFromSmarts('[A]').GetAtomWithIdx(0)
def MotifToTemplate(smiles):
    mol = Chem.MolFromSmiles(smiles)
    amap = list(i for i in range(mol.GetNumAtoms()) if mol.GetAtomWithIdx(i).GetSymbol() != '*')
    motif = Chem.RWMol(Chem.MolFromSmarts(''))
    for aix in amap:
        if mol.GetAtomWithIdx(aix).GetIsAromatic():
            motif.AddAtom(aromaticA)
        else:
            motif.AddAtom(aphalicA)
    for i, j in combinations(amap, 2):
        bond = mol.GetBondBetweenAtoms(i, j)
        if bond:
            motif.AddBond(amap.index(i),
               amap.index(j), Chem.BondType.UNSPECIFIED)
    Chem.SanitizeMol(motif)
    return Chem.MolToSmarts(motif)
```
However, I find that Chem.MolToSmarts cannot do canonicalization, example here:
```python3
MotifToTemplate("c1ccc2[nH]ccc2c1") # gives "a1aaa2aaaa2a1"
MotifToTemplate("c1cc2ccccc2[nH]1") # gives "a1aa2aaaaa2a1"
```
Same molecule can give different skeleton SMARTS, let alone different molecules (even with same topology)! Any suggestions on how to get canonical SMARTS, or is there any other suggestions for converting a molecule to this type of "skeleton"?
Many thanks!!
Jocelyn

**Comment by bp-kelley** [comment]
This probably won't work all the time, but you could canonicalize your smiles on input, i.e.:

```
def MotifToTemplate(smiles):
    mol = Chem.MolFromSmiles(Chem.CanonSmiles(smiles))
   ...
```

```
print(MotifToTemplate("c1ccc2[nH]ccc2c1")) # gives "a1aaa2aaaa2a1"
print(MotifToTemplate("c1cc2ccccc2[nH]1")) # gives "a1aa2aaaaa2a1"
```

```
a1aaa2aaaa2a1
a1aaa2aaaa2a1
```

**Reply by HelloJocelynLu** [reply, marked as answer]
Thank you for your help! However sometime molecules with the same topology does not give the same result. For example:
<img width="235" alt="image" src="https://github.com/rdkit/rdkit/assets/34988596/1ae493be-3eba-4772-bbee-f9ed9458966a">

```pyhton
Chem.CanonSmiles("c1ncc2[nH]ccc2c1") # gives c1cc2cc[nH]c2cn1, so a1aa2aaaa2aa1
Chem.CanonSmiles("c1ncc2[nH]ccc2c1") # gives c1ccc2[nH]ccc2c1, so a1aaa2aaaa2a1
```

Fortunately, Inspired by your method, I've come up with an idea:
```python
def MotifToTemplate(smiles):
        mol = Chem.MolFromSmiles(smiles)
        fake_mol = Chem.ReplaceSubstructs(smiles, Chem.MolFromSmarts("*"), Chem.MolFromSmiles("C"), replaceAll=True)[0]
        fake_smiles = Chem.MolToSmiles(fake_mol)
        ...
```
Thanks Kelley!!

**Comment by iuhgnor** [comment]
`GetSubstructMatch` can be used to determines whether two SMARTS are the same. 

```python
def is_same_topology(sma1, sma2):
    patt1 = Chem.MolFromSmarts(sma1)
    patt2 = Chem.MolFromSmarts(sma2)
    mol1 = Chem.MolFromSmarts(sma1)
    mol2 = Chem.MolFromSmarts(sma2)
    if mol1.GetSubstructMatch(patt2) and mol2.GetSubstructMatch(patt1):
        return True
    else:
        return False

sma1 = 'a1aaa2aaaa2a1'
sma2 = 'a1aa2aaaaa2a1'

is_same_topology(sma1, sma2)
```

```txt
True
```

**Reply by HelloJocelynLu** [reply]
Thank you for your response. Yes I can do that, but if I have a lot of `sma`, the pair-wise comparison becomes a problem. Luckily I've find a solution. Glad to have you in this discussion!
